### DISCUSSION TITLE
Handling null parent element

### DISCUSSION THREAD

**Discussion by joel-garcia** [colinhacks/zod#3258] [Q&A] [answered]
What is the proper way to structure a ZodObject to handle an instance where an API returns a null value for a parent element which contains a single child element? The format of the data comes from the API and I have to use Prisma composite types to push the data into MongoDB,
```
thisSchema= z.object({
    (...other elements)
    parentElement: z.object({
        childElement: z.string().url(),
    })
})
```
There are some cases where the API will return data where the parentElement is null. This is acceptable in my Prisma schema, but I am having trouble setting up the ZodObject to validate correctly.

Do I need to mark both the parentElement and/or childElement as optional/nullish somehow? When I try just adding optional/nullish to the childElement I get a TypeScript error telling me:
```
The types of  parentElement.childElement are incompatible between these types.
Type  string | null | undefined  is not assignable to type  string 
Type  undefined  is not assignable to type  string 
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.object( {
    parentElement: z.object( {
        childElement: z.string().url(),
    } ).nullish()
} )

console.log(
    schema.parse( {
        parentElement: {
            childElement: 'https://example.com'
        }
    } )
)
// {
//     parentElement: {
//         childElement: "https://example.com",
//     },
// }

console.log(
    schema.parse( {
        parentElement: null
    } )
)
// {
//     parentElement: null,
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by joel-garcia** [reply]
Adding nullish() to the parent object allows my code to run, but still generates a TypeScript error (TS2322). 
```
Types of property parentElement are incompatible.
Type
{
    childElement: string;
} | null | undefined
is not assignable to type ParentCompositeChildCreateInput
Type undefined is not assignable to type ParentCompositeChildCreateInput
```
Which seems to say my Prisma schema is unable to accept a potential value of undefined. The Prisma schema is currently defined as:
```
type ParentComposite{
  parentElement    ParentCompositeChild?
}

type ParentCompositeChild{
  childElement    String?
}
```
I'm not sure this is something I need to be overly concerned with, since the code functions correctly with the addition of nullish() on the parentElement.

**Comment by JacobWeisenburger** [comment]
```ts
const schema = z.object( {
    parentElement: z.object( {
        childElement: z.string().url(),
    } ).nullable()
} )

console.log(
    schema.parse( {
        parentElement: {
            childElement: 'https://example.com'
        }
    } )
)
// {
//     parentElement: {
//         childElement: "https://example.com",
//     },
// }

console.log(
    schema.parse( {
        parentElement: null
    } )
)
// {
//     parentElement: null,
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by joel-garcia** [reply]
I also tried using nullable() instead before my last reply, but that merely changes the error to:
```
Type null is not assignable...
```
So, it seems it doesn't like either null or undefined.

**Reply by JacobWeisenburger** [reply, marked as answer]
I don't really know Prisma, so you'll have to figure out the type that Prisma is trying to give you.
